Config flow, Update coordinator and device info for SAJ integration

sanyatuning opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Reopened PR #41198
SAJ integration can be configured thru the UI.
Add a device to the device registry to connect entities.
Use DataUpdateCoordinator instead of custom logic
Log deprication warning when configured with yaml.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #40533
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16677

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository. #48337 #48170

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

corneyl commented on homeassistant/components/saj/config_flow.py:
```diff
@@ -0,0 +1,74 @@
+            host = user_input["host"]
+
+            await self.async_set_unique_id(host)
+            self._abort_if_unique_id_configured()
```
Isn't it better to request the serial number from the inverter and use that as unique id? The hostname might be just a setting on the device.

corneyl commented on homeassistant/components/saj/__init__.py:
```diff
@@ -1 +1,24 @@
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the saj component."""
+    hass.data[DOMAIN] = {}
```
`hass.data[DOMAIN] = {}` is initialised but never used. You can use it to save a instance of `SAJInverter` which then can be used in the platforms.

corneyl commented on homeassistant/components/saj/sensor.py:
```diff
@@ -56,53 +58,97 @@
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
```
`async_add_entities` has no type, maybe good to add?

corneyl commented on homeassistant/components/saj/sensor.py:
```diff
@@ -120,61 +166,59 @@ async def async_saj():
     """Representation of a SAJ sensor."""
 
-    def __init__(self, serialnumber, pysaj_sensor, inverter_name=None):
+    def __init__(self, inverter: SAJInverter, pysaj_sensor):
```
Would be good to add a type hint to `pysaj_sensor` imo.

corneyl reviewed (commented):
Would be nice to have some tests for the sensor platform as well!

sanyatuning commented:
@fredericvl can you run some tests for me or a review?

sanyatuning commented:
reviewed: #48337 #48170

sanyatuning commented:
@corneyl can you review the tests added?

sanyatuning commented:
@frenck sorry for the interrupt, the maintainer didn't respond, can you check this?

fredericvl commented on homeassistant/components/saj/sensor.py:
```diff
@@ -119,61 +177,59 @@ async def async_saj():
         return values
 
-    def start_update_interval(event):
+    def start_update_interval(self, _):
```
When adding the integration or restarting HA the sensors stay in an unknown state until the interval time has been reached.
It would be nice if the sensors immediately have the retrieved values and from then on schedules updates every X seconds.

fredericvl reviewed (commented):
Nice work @sanyatuning 👍 
I only have two comments for some improvements.
It would be nice if you could have a look at those.

sanyatuning commented on homeassistant/components/saj/sensor.py:
```diff
@@ -15,25 +17,26 @@
 
-INVERTER_TYPES = ["ethernet", "wifi"]
+# seconds
+MIN_INTERVAL = 30
```
Thanks for the feedback. Now you can set it via configuration.yaml or the UI.

sanyatuning commented:
@fredericvl can you check this again?
I added an option to override the default interval.
And I refactored the polling logic to use DataUpdateCoordinator.

fredericvl commented:
> @fredericvl can you check this again?
> I added an option to override the default interval.
> And I refactored the polling logic to use DataUpdateCoordinator.

Looks nice and it works at my end!
I just noticed that when changing the interval that it does not reflect realtime and you have to restart HA.
According to the documentation this should be possible: https://developers.home-assistant.io/docs/config_entries_options_flow_handler/#signal-updates
That would be even more user friendly.
Would you mind having a look at that?

sanyatuning commented:
> Looks nice and it works at my end!
> I just noticed that when changing the interval that it does not reflect realtime and you have to restart HA.
> According to the documentation this should be possible: https://developers.home-assistant.io/docs/config_entries_options_flow_handler/#signal-updates
> That would be even more user friendly.
> Would you mind having a look at that?

try it now

sanyatuning commented:
@fredericvl ping

frenck commented on homeassistant/components/saj/__init__.py:
```diff
@@ -1 +1,49 @@
+PLATFORMS = ["sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the saj component."""
+    return True
+
+
```
This isn't needed anymore.

```suggestion
```

frenck commented on homeassistant/components/saj/__init__.py:
```diff
@@ -1 +1,49 @@
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
```
Missing return type

```suggestion
async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```

frenck commented on homeassistant/components/saj/__init__.py:
```diff
@@ -1 +1,49 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up saj from a config entry."""
+
+    config = entry.data.copy()
+    config.update(entry.options)
+    inverter = SAJInverter(config)
```
We should explicitly create the configuration and not just pass the dictionary as it. (which would remove the need for things like making a copy as well)

frenck replied:
I'm missing some kind of connection testing here?
The config flow does that, but on entry setup there is nothing?
Should it check if it has auth for example?

sanyatuning replied:
I removed the connection check because these inverters were only available in the daytime.
Now, if you restart HA after sunset it became available the next day. 
If we had another method I'm open to that.

frenck replied:
When a connection fails, it can raise a ConfigEntryNotReady, HA will keep retrying.

sanyatuning replied:
I will try this.

sanyatuning replied:
It works, thanks

frenck commented on homeassistant/components/saj/__init__.py:
```diff
@@ -1 +1,49 @@
+    config = entry.data.copy()
+    config.update(entry.options)
+    inverter = SAJInverter(config)
+    undo_listener = entry.add_update_listener(update_listener)
```
```suggestion
	entry.async_on_unload(entry.add_update_listener(update_listener))
```

frenck commented on homeassistant/components/saj/__init__.py:
```diff
@@ -1 +1,49 @@
+    undo_listener = entry.add_update_listener(update_listener)
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        DATA_INVERTER: inverter,
+        DATA_UNDO_UPDATE_LISTENER: undo_listener,
+    }
```
```suggestion
    hass.data[DOMAIN][entry.entry_id] = inverter
```

frenck commented on homeassistant/components/saj/__init__.py:
```diff
@@ -1 +1,49 @@
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+
+from .const import DATA_INVERTER, DATA_UNDO_UPDATE_LISTENER, DOMAIN
```
```suggestion
from .const import DOMAIN
```

frenck commented on homeassistant/components/saj/__init__.py:
```diff
@@ -1 +1,49 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    if unload_ok:
+        hass.data[DOMAIN][entry.entry_id][DATA_UNDO_UPDATE_LISTENER]()
+        hass.data[DOMAIN].pop(entry.entry_id)
```
```suggestion
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        hass.data[DOMAIN].pop(entry.entry_id)
```

frenck commented on homeassistant/components/saj/__init__.py:
```diff
@@ -1 +1,49 @@
+
+async def update_listener(hass: HomeAssistant, entry: ConfigEntry):
+    """Handle options update."""
+    inverter: SAJInverter = hass.data[DOMAIN][entry.entry_id][DATA_INVERTER]
+    inverter.update_options(entry.options)
```
we can also reload the entry instead?

sanyatuning replied:
removed

frenck commented on homeassistant/components/saj/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+)
+from homeassistant.core import callback
+
+from .const import (  # pylint:disable=unused-import
```
Should not be needed to disable pylint here

```suggestion
from .const import (
```

frenck commented on homeassistant/components/saj/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+    """Handle a config flow for SAJ."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
```
This isn't used anymore

```suggestion
```

frenck commented on homeassistant/components/saj/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry: ConfigEntry):
```
Either provide a full signature or none :)

```suggestion
    def async_get_options_flow(config_entry: ConfigEntry) -> OptionsFlowHandler:
```

frenck commented on homeassistant/components/saj/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+        errors = {}
+        if user_input is not None:
+            try:
+                inverter = SAJInverter(user_input)
```
Don't pass the user_input directly, build the parameters it needs.

frenck commented on homeassistant/components/saj/config_flow.py:
```diff
@@ -0,0 +1,125 @@
+            data_schema=(
+                vol.Schema(
+                    {
+                        vol.Optional(
+                            CONF_SCAN_INTERVAL,
+                            default=self.config_entry.options.get(
+                                CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
+                            ),
+                        ): vol.All(
+                            vol.Coerce(int),
+                            vol.Clamp(min=MIN_SCAN_INTERVAL, max=MAX_SCAN_INTERVAL),
+                        ),
+                    }
```
We no longer need a scan interval to be configurable.
Please set a sane default, user can manually disable polling and update using automations on their own schedule. This is a generic feature of Home Assistant since 2021.6.

sanyatuning replied:
30s is OK?

frenck commented on homeassistant/components/saj/sensor.py:
```diff
@@ -48,139 +58,161 @@
     {
         vol.Required(CONF_HOST): cv.string,
         vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(CONF_SCAN_INTERVAL): cv.positive_int,
```
We no longer accept changes to the YAML configuration schema. Please revert this change.

frenck commented on homeassistant/components/saj/sensor.py:
```diff
@@ -48,139 +58,161 @@
-            sensor.async_update_values(unknown_state=state_unknown)
 
-        return values
+class SAJInverter:
```
This is not part of the sensor anymore, I suggest moving this out of the sensor file.

This class could be extending the DataUpdateCoordinator (instead of wrapping it).

frenck commented on homeassistant/components/saj/sensor.py:
```diff
@@ -48,139 +58,161 @@
+        """Return enabled sensors keys."""
+        return [s.key for s in self._sensor_def if s.enabled]
 
+    @property
+    def available(self) -> bool:
+        """Return True if device is available."""
+        return self.coordinator and self.coordinator.last_update_success
```
This is handled on the entity level by the CoordinatorEntity, it isn't used in this class.

frenck commented on homeassistant/components/saj/sensor.py:
```diff
@@ -48,139 +58,161 @@
+        "Loading SAJ Solar inverter integration via yaml is deprecated. "
+        "Please remove it from your configuration"
+    )
+    inverter = SAJInverter(config)
+    await inverter.setup(hass, async_add_entities)
```
We should not instantiate from config once the config flow has been in place. Instead, trigger the configuration flow to import the old YAML configuration into a config entry.

frenck reviewed (changes requested):
Hi there @sanyatuning 👋 

Thanks for opening a PR. A good start!

I think there is some entanglement between the old style and the new style you are trying to introduce. It would need some refactoring to get most of the old logic out of the sensor platform (as it should be handled on the integration level now).

This is especially visible when looking at the integration init, which actually now pulls in / diverts to partially previous code in the sensor. This makes the code hard to follow and unneeded complex.

I hope the comments can help you started. Let me know if I can be at help.

../Frenck

frenck commented on homeassistant/components/saj/sensor.py:
```diff
@@ -178,15 +101,15 @@ def __init__(self, inverter: SAJInverter, pysaj_sensor: pysaj.Sensor) -> None:
     def state(self):
         """Return the state of the sensor."""
-        if not self._inverter.available:
+        if not self.coordinator.last_update_success:
```
When using the coordinator, this check isn't needed.
`state()` isn't called when the entity isn't unavailable.

sanyatuning replied:
some entity like "today_yield" is still available after sundown when the inverter is not reachable (assumed state?)
pls see comment in lines below

sanyatuning replied:
@frenck ping

sanyatuning commented:
Please some I opened this PR in February

frenck commented on homeassistant/components/saj/coordinator.py:
```diff
@@ -0,0 +1,91 @@
+        """Try to connect to the inverter."""
+        try:
+            done = await self._saj.read(self._sensor_def)
+            if done:
+                return
```
These 2 lines cannot raise and should be outside of the try/except scope.

frenck commented on homeassistant/components/saj/coordinator.py:
```diff
@@ -0,0 +1,91 @@
+
+    async def setup(self, async_add_entities: Callable):
+        """Add sensors to Core and get first state."""
+        # pylint: disable=import-outside-toplevel
+        from .sensor import SAJSensor
```
uh? Why do we need to import outside of the top level?

sanyatuning replied:
Was a circular dependency error here. Coordinator vs the sensor. Any suggestion?

frenck replied:
Yes, move it into the sensor platform :)

frenck commented on homeassistant/components/saj/coordinator.py:
```diff
@@ -0,0 +1,91 @@
+        # pylint: disable=import-outside-toplevel
+        from .sensor import SAJSensor
+
+        await self.async_refresh()
+        async_add_entities(
+            SAJSensor(self, sensor) for sensor in self._sensor_def if sensor.enabled
+        )
```
This belongs in the sensor platform

frenck commented on homeassistant/components/saj/coordinator.py:
```diff
@@ -0,0 +1,91 @@
+
+    async def update(self):
+        """Fetch data from Inverter."""
+        done = await self._saj.read(self._sensor_def)
```
Don't we have to try/except this method? What can it raise?

sanyatuning replied:
I think this is fine.
https://github.com/fredericvl/pysaj/blob/master/pysaj/__init__.py#L133

frenck commented on homeassistant/components/saj/sensor.py:
```diff
@@ -63,119 +60,39 @@
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: Callable
+):  # pragma: no cover
```
```suggestion
):
```

frenck replied:
Additionally, return type is missing.

frenck commented on homeassistant/components/saj/sensor.py:
```diff
@@ -63,119 +60,39 @@
+
+async def async_setup_platform(
+    hass: HomeAssistant, config, async_add_entities: Callable, discovery_info=None
+):  # pragma: no cover
```
```suggestion
):
```

frenck replied:
Additionally, the return type is missing.

frenck commented on homeassistant/components/saj/sensor.py:
```diff
@@ -63,119 +60,39 @@
+
+
+async def async_setup_platform(
+    hass: HomeAssistant, config, async_add_entities: Callable, discovery_info=None
```
Please fully type methods (instead of partial typing).

frenck commented on homeassistant/components/saj/sensor.py:
```diff
@@ -185,15 +102,31 @@ def __init__(self, serialnumber, pysaj_sensor, inverter_name=None):
+        """
+        Return True if device is available.
+
+        Note: SAJ inverters are powered by DC via solar panels and thus are
+        offline after the sun has set, but we keep some sensor available.
```
So, if I understand it correctly, it keeps the entity available, while the device is actually already offline?

sanyatuning replied:
Yes, some sensor

frenck replied:
But that sounds incorrect, if a device is unavailable, the entities should reflect that.

jwillemsen replied:
it is helpful to see the today_time, today_yield, total_co2_reduced, total_time, and total_yield even when the saj is offline, the current_power/temperature are valid when the saj is online

frenck replied:
> it is helpful to see the today_time, today_yield, total_co2_reduced, total_time, and total_yield even when the saj is offline, the current_power/temperature are valid when the saj is online

Instead, those sensors could be exempted from being unavailable maybe?

sanyatuning replied:
This is what the code does. I don't get it. :)

frenck commented on homeassistant/components/saj/strings.json:
```diff
@@ -0,0 +1,33 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "Connect to the device",
```
Does not really add something here

```suggestion
```

frenck commented on homeassistant/components/saj/strings.json:
```diff
@@ -0,0 +1,33 @@
+      "user": {
+        "title": "Connect to the device",
+        "data": {
+          "host": "IP address of the inverter (should have static IP address)",
```
```suggestion
          "host": "[%key:common::config_flow::data::host%]",
```

frenck commented on homeassistant/components/saj/strings.json:
```diff
@@ -0,0 +1,33 @@
+          "host": "IP address of the inverter (should have static IP address)",
+          "type": "Connection type",
+          "name": "Inverter name (optional)",
+          "username": "Username (optional)",
```
```suggestion
          "username": "[%key:common::config_flow::data::username%]",
```

frenck commented on homeassistant/components/saj/strings.json:
```diff
@@ -0,0 +1,33 @@
+          "type": "Connection type",
+          "name": "Inverter name (optional)",
+          "username": "Username (optional)",
+          "password": "Password (optional)"
```
```suggestion
          "password": "[%key:common::config_flow::data::password%]"
```

frenck commented on homeassistant/components/saj/strings.json:
```diff
@@ -0,0 +1,33 @@
+        "data": {
+          "host": "IP address of the inverter (should have static IP address)",
+          "type": "Connection type",
+          "name": "Inverter name (optional)",
```
```suggestion
```

frenck commented on homeassistant/components/saj/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+                    vol.Required(CONF_TYPE, default=user_input.get(CONF_TYPE)): vol.In(
+                        INVERTER_TYPES
+                    ),
+                    vol.Optional(CONF_NAME, default=user_input.get(CONF_NAME, "")): str,
```
We shouldn't ask for the name, instead, use a default name or maybe an IP address.
The end-user can always decide to rename the device if they like

frenck commented on homeassistant/components/saj/strings.json:
```diff
@@ -0,0 +1,33 @@
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
+    }
+  },
+  "options": {
+    "step": {
+      "init": {
+        "data": {
+          "scan_interval": "Seconds between updates"
+        }
+      }
+    }
+  }
```
```suggestion
  }
```

sanyatuning commented on homeassistant/components/saj/coordinator.py:
```diff
@@ -0,0 +1,91 @@
+            update_interval=UPDATE_INTERVAL,
+            update_method=self.update,
+        )
+        self.last_update_success = False
```
The default is true.
https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/update_coordinator.py#L65

sanyatuning commented on homeassistant/components/saj/coordinator.py:
```diff
@@ -0,0 +1,91 @@
+            update_method=self.update,
+        )
+        self.last_update_success = False
+        self._saj = saj or _init_pysaj(wifi, host, username, password)
```
saj is passed in tests

frenck commented:
Hmm, this one seems stuck (totally our bad, sorry about that @sanyatuning 🙏) Could you rebase this PR onto the latest `dev` branch? Thanks!

sanyatuning commented:
Ok I will

vcc-sanya committed:
Add config flow and device info to SAJ integration

vcc-sanya committed:
skip connect on config entry import

vcc-sanya committed:
add default values for config entry import

vcc-sanya committed:
test fixes

vcc-sanya committed:
test fixes

vcc-sanya committed:
test with minimal yaml config

sanyatuning commented:
@frenck can you review this?

sverleysen commented:
@frenck @sanyatuning, any idea when this PR will be merged? It will fix an issue I have: #40533

frenck commented:
Nope, this is not a schedule system.

sverleysen commented:
Oh, it was mentioned in the PR description, so I though/hoped it would fix it.
